hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/MultiVersionConcurrencyControl.java

```diff
@@ -36,13 +36,13 @@
  */
 @InterfaceAudience.Private
 public class MultiVersionConcurrencyControl {
-  private static final Logger LOG = LoggerFactory.getLogger(MultiVersionConcurrencyControl.class);
   private static final long READPOINT_ADVANCE_WAIT_TIME = 10L;
 
   final String regionName;
   final AtomicLong readPoint = new AtomicLong(0);
   final AtomicLong writePoint = new AtomicLong(0);
-  private final Object readWaiters = new Object();
+
+  private final ThreadLocal<WriteEntry> cachedWriteEntries;
```

Comment:
Is it always safe to use ThreadLocal here? What if users just use ASYNC_WAL or even NO_WAL and return without waiting for MVCC to move？

Reply:
Let me check the multi-thread safety more carefully. 
User will wait for MVCC to move when use  `ASYNC_WAL` or even `NO_WAL`
`ASYNC_WAL` only  effect phase of WAL sync which contained in the period from `WriteEntry` generate to  `MVCC` move, `NO_WAL` will generate `WriteEntry` while write data to memstore

Reply:
Needs comment explaining lifeccycle of cachedWriteEntries. Why is it plural? Should it be cachedWriteEntry? Why not declare and initialize here rather then declare and then initialize in the constructor?

Reply:
I like the idea of saving on allocation of a WriteEntry.

The @Apache9 questions are good ones.